### PR TITLE
🐛 ✨ Is353/meta func command and fixes

### PR DESCRIPTION
**Pull request by pcrespov**
## **TL;DR** 

Now we can override the default command ``run``on any computational service in the image annotation labels. For instance:
```yaml
io.simcore.type: '{"type": "computational"}'     
simcore.service.settings: '[{"name": "ContainerSpec", "type": "ContainerSpec", "value": {"Command": ["prog", "arg1"]}}]'
#...
``` 
will start the service container using CLI ``prog arg1`` instead of the default ``run``.

## What do these changes do?

We need a way to override the default container CMD (i.e. ``["run"]``) that is used to run a computational service.  This shall be specified as part of the  ``simcore.service.settings`` annotation label in the service image. The model used in that entry is roughly a list of structured data defined in ``models_library.service_settings_labels.SimcoreServiceSettingLabelEntry`` . This is essentially a generic container that was designed to allocate values used in parts of the request body for https://docs.docker.com/engine/api/v1.41/#operation/ServiceCreate.  

This PR adds a bit more structure (and doc) to this design and extends to allow overriding ``TaskTemplate- > ContainerSpec -> Command`` (see ``ContainerSpec``)

This information is transmitted throughout several services: The service image labels in docker ``registry``  --> ( ``registry_proxy``) -->  ``director``  --> exposed via  ``GET /service_extras/{service_key}/{service_version}`` ) --> director-v2  -->  ``Image`` models 

This change will allow running services like [``osparc-function-services``](https://github.com/pcrespov/osparc-function-services) with ``ContainerSpec``s.

NOTE:  in director-v2, ``ServiceExtras``  and GET ``/{service_key:path}/{service_version}/extras`` is a redirection to ``director``'s API

#### ``models_library``:
   -  ✨ new ``service_settings_labels.ContainerSpec`` model
   -  ♻️ ``test__models_examples::test_all_module_model_examples``  automatically tests all examples in ``models_library`` against model schemas.

#### ``director`` service:
- ✨``reverse_proxy`` can parse ``container_spec`` from ``simcore.service.settings`` image annotation label's values.

#### ``director-v2`` service:
- ✨``Image`` adds ``command`` that can be overridden by ``container_spec.command`` above.

#### ``servie-integration-library``:
- ✨ enhances validation of ``osparc_image_specs``, specifically ``SettingsItem``

#### Other
- ⚗️ ``scripts/demo-meta``: script to retrieve meta-study results via API
- 🐛 ``webserver``: minor fix on definition of meta-study
-♻️ minor cleanup and doc on logging and models

## Related issue/s

- ITISFoundation/osparc-issues#353

## How to test

- tests for all modules mentioned above

## checklist

- [x] Openapi changes? ``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Database migration script? ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
- [x] Unit tests for the changes exist
- [ ] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

**Review by mguidon — COMMENTED** [review]
Overriding the COMMAND is very useful!

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 1]
```diff
@@ -16,10 +16,39 @@ class _BaseConfig:
     keep_untouched = (cached_property,)
 
 
+class ContainerSpec(BaseModel):
+    """Implements entries can can be overriden for https://docs.docker.com/engine/api/v1.41/#operation/ServiceCreate
```
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:20]:
```suggestion
    """Implements entries that can be overriden for https://docs.docker.com/engine/api/v1.41/#operation/ServiceCreate
```

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 2]
```diff
@@ -16,10 +16,39 @@ class _BaseConfig:
...
+    request body: TaskTemplate -> ContainerSpec
+    """
+
+    command: list[str] = Field(
+        alias="Command",
+        description="The command to be run in the image. If None, it will not override default 'run'",
```
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:26]: Q: should we not have actually "run" as the default?

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 3]
```diff
@@ -16,10 +16,39 @@ class _BaseConfig:
...
+            "examples": [
+                {"Command": ["executable"]},
+                {"Command": ["executable", "subcommand"]},
+                {"Command": ["ofs", "linear-regression"]},
+            ]
+        }
+
+
 class SimcoreServiceSettingLabelEntry(BaseModel):
+    """These value is used to build the request body of https://docs.docker.com/engine/api/v1.41/#operation/ServiceCreate
```
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:43]:
```suggestion
    """These value are used to build the request body of https://docs.docker.com/engine/api/v1.41/#operation/ServiceCreate
```

Comment (sanderegg) [reply packages/models-library/src/models_library/service_settings_labels.py:43]: well, if one day we move to kubernetes, this comment might not be really necessary here. but that is ok.

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 4]
```diff
@@ -38,7 +67,13 @@ class Config(_BaseConfig):
                     "type": "string",
                     "value": ["node.platform.os == linux"],
                 },
-                # resources
+                # SEE service_settings_labels.py::ContainerSpec
+                {
+                    "name": "ContainerSpec",
+                    "type": "ContainerSpec",
```
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:73]: not sure I remember very well what I tried to do, but should the type not be "Command"? This is what defined the array right?

File: packages/models-library/tests/test__models_examples.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                # NOTE: this is part of utils.misc and is tested here
+                examples = extract_examples(model_cls)
+                for index, example in enumerate(examples):
+                    yield (name, model_cls, index, example)
+
+
+@pytest.mark.parametrize(
+    "class_name, model_cls, example_index, test_example", iter_model_cls_examples()
+)
+def test_all_module_model_examples(
```
Comment (sanderegg) [inline packages/models-library/tests/test__models_examples.py:46]: Q: hmm this is a rewrite of test_utils_misc right? is there something specific to know about it? or is it mostly just naming?

File: packages/service-integration/src/service_integration/osparc_config.py [hunk 6]
```diff
@@ -148,6 +154,14 @@ class SettingsItem(BaseModel):
         description="The value of the service setting (shall follow Docker REST API scheme for services",
```
Comment (sanderegg) [inline packages/service-integration/src/service_integration/osparc_config.py:154]: Reading this now, I am not sure I understand why it's rewritten.
Also, the new concept is to run containers, not services (at least from the sidecars) so the description is actually incorrect. sorry for that. see https://docs.docker.com/engine/api/v1.41/#operation/ContainerCreate

File: services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py [hunk 7]
```diff
@@ -24,16 +24,31 @@
...
     name: str = Field(..., regex=KEY_RE)
     tag: str = Field(..., regex=VERSION_RE)
-    requires_gpu: Optional[bool] = Field(None, deprecated=True)
-    requires_mpi: Optional[bool] = Field(None, deprecated=True)
+
+    requires_gpu: Optional[bool] = Field(
+        None, deprecated=True, description="Use instead node_requirements"
+    )
+    requires_mpi: Optional[bool] = Field(
+        None, deprecated=True, description="Use instead node_requirements"
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py:32]: is there any necessity of keeping both?

Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py:32]: also what is the meaning of having node_requirements defaults to None? it should at least have CPU right? before since there was no requires_cpu entry it was implicit. now it should be explicit.

File: services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py [hunk 8]
```diff
@@ -24,16 +24,31 @@
...
+    )
 
     @validator("node_requirements", pre=True, always=True)
     @classmethod
     def migrate_from_requirements(cls, v, values):
         if v is None:
+            # NOTE: 'node_requirements' field's default=None although is NOT declared as nullable.
+            # Then this validator with `pre=True, always=True` is used to create a default
+            # based on that accounts for an old version.
+            # This strategy guarantees backwards compatibility
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py:51]: ok I see your point. I guess we'll iterate over this again with the personalized resources case again.

File: services/director-v2/src/simcore_service_director_v2/models/schemas/services.py [hunk 9]
```diff
@@ -57,7 +58,8 @@ class Config:
 
 class ServiceExtras(BaseModel):
     node_requirements: NodeRequirements
-    service_build_details: Optional[ServiceBuildDetails]
+    service_build_details: Optional[ServiceBuildDetails] = None
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/models/schemas/services.py:61]: hmm, I guess I'll need to review all this Options... this must come from lack of knowledge of pydantic at the time. cause this is probably only optional for old services

File: services/director-v2/src/simcore_service_director_v2/modules/dask_client.py [hunk 10]
```diff
@@ -310,7 +311,7 @@ def _comp_sidecar_fct(
                     input_data=input_data,
                     output_data_keys=output_data_keys,
                     log_file_url=log_file_url,
-                    command=["run"],
+                    command=node_image.command,
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dask_client.py:314]: you can also add in the  dask-sidecar code, that it prints the command in the logs as well. could be helpful for debugging

File: services/director/src/simcore_service_director/registry_proxy.py [hunk 11]
```diff
@@ -375,69 +375,91 @@ def get_service_last_names(image_key: str) -> str:
...
-            .get("Reservations", {})
-            .get("GenericResources", [])
-        ):
-            if element.get("DiscreteResourceSpec", {}).get("Kind") == kind_name:
-                return True
-        return False
+    labels = await get_image_labels(app, image_key, image_tag)
+    logger.debug("Compiling service extras from labels %s", pformat(labels))
 
+    # check if the service requires GPU support
```
Comment (sanderegg) [inline services/director/src/simcore_service_director/registry_proxy.py:408]: are you sure about that comment?

**Review by sanderegg — COMMENTED** [review]
some early / late comments... we can discuss tomorrow. great work!

File: packages/models-library/src/models_library/service_settings_labels.py (hunk 2 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/service_settings_labels.py:26]: I prefer to do that when I get to the last model's interface i.e. ``Image``. 
At this point, it is part of the ``models_library`` and it is used only as a mean to override.

File: packages/models-library/tests/test__models_examples.py (hunk 5 above)
Comment (pcrespov) [reply packages/models-library/tests/test__models_examples.py:46]: now the parametrization distinguishes the ``test_example`` as well. Before they were all inside.
This is a very neat way to test all examples in all models defined in ``models_library``. 

I will probably create a helper in a separate PR to do that in all packages.

File: packages/service-integration/src/service_integration/osparc_config.py (hunk 6 above)
Comment (pcrespov) [reply packages/service-integration/src/service_integration/osparc_config.py:154]: LEt's double check that tomorrow

File: services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py (hunk 7 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py:32]: Those were already there and marked as deprecated. 

The validator below provides an automatic migration and  guarantees backward compatibility. Check the comment in the validator ``def migrate_*``

File: services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py (hunk 8 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py:51]: yes.

File: services/director-v2/src/simcore_service_director_v2/modules/dask_client.py (hunk 10 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/modules/dask_client.py:314]: You can find that some lines below ...

File: services/director/src/simcore_service_director/registry_proxy.py (hunk 11 above)
Comment (pcrespov) [reply services/director/src/simcore_service_director/registry_proxy.py:408]: That comment was already there and I guess I missed it.

In any case, this module needs a complete refactoring. I just patched it up to make it work. A clean sdk to access the registry shall replace this

File: packages/models-library/src/models_library/service_settings_labels.py (hunk 4 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/service_settings_labels.py:73]: no, the type is ``ContainerSpec`` and has a field called ``Command``
```yaml
io.simcore.type: '{"type": "computational"}'     
simcore.service.settings: '[{"name": "ContainerSpec", "type": "ContainerSpec", "value": {"Command": ["prog", "arg1"]}}]'
#...
```

File: packages/models-library/tests/test__models_examples.py (hunk 5 above)
Comment (sanderegg) [reply packages/models-library/tests/test__models_examples.py:46]: what about using hypothesis?

File: packages/models-library/src/models_library/service_settings_labels.py (hunk 4 above)
Comment (sanderegg) [reply packages/models-library/src/models_library/service_settings_labels.py:73]: let's discuss that one

File: services/director-v2/src/simcore_service_director_v2/models/schemas/services.py (hunk 9 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/models/schemas/services.py:61]: and that code is probably super old and when we did not know half of what pydantic can do. and that we have services that did not have the build details...

File: services/director-v2/src/simcore_service_director_v2/modules/dask_client.py (hunk 10 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dask_client.py:314]: indeed I did not see it modified, that should be in there:
https://github.com/sanderegg/osparc-simcore/blob/master/services/dask-sidecar/src/simcore_service_dask_sidecar/tasks.py#L1

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 12]
```diff
@@ -16,10 +16,39 @@ class _BaseConfig:
...
+    """Implements entries that can be overriden for https://docs.docker.com/engine/api/v1.41/#operation/ServiceCreate
+    request body: TaskTemplate -> ContainerSpec
+    """
+
+    command: list[str] = Field(
+        alias="Command",
+        description="Used to override the container's command",
+        # NOTE: currently constraint to our use cases. Might mitigate some security issues.
+        min_items=1,
+        max_items=2,
```
Comment (GitHK) [inline packages/models-library/src/models_library/service_settings_labels.py:29]: Even with the blow examples It's tricky to understand why only max 2 arguments. Usually users have more complex cli.

File: packages/service-integration/src/service_integration/osparc_config.py (hunk 6 above)
Comment (GitHK) [reply packages/service-integration/src/service_integration/osparc_config.py:154]: Please remember this is also used for validating `dynamic-services` as well.

File: services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py [hunk 13]
```diff
@@ -78,23 +103,23 @@ class Config:
...
-    end: Optional[datetime]
+    start: Optional[datetime] = Field(default=None)
+    end: Optional[datetime] = Field(default=None)
     state: RunningState
-    task_id: Optional[PositiveInt]
+    task_id: Optional[PositiveInt] = Field(default=None)
     internal_id: PositiveInt
     node_class: NodeClass
-    errors: Optional[list[ErrorDict]]
+    errors: Optional[list[ErrorDict]] = Field(default=None)
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py:122]: Why add all these default None values? I guess it does not have an impact on the current codebase

**Review by GitHK — APPROVED** [review]
👍

File: packages/models-library/src/models_library/service_settings_labels.py (hunk 12 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/service_settings_labels.py:29]: As the note suggests ... I will constraint for the moment to our use case

File: services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py (hunk 13 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/models/domains/comp_tasks.py:122]: It has an impact with the static type check system
We mentioned some PRs ago that pylance will complain if the defaults where not explicitly set as ``Field(default=** )``

Here [it defaults to None](https://github.com/ITISFoundation/osparc-simcore/blob/bee16a3398a8ee5ed6fe99f5a17cbbfaec7d75f7/packages/settings-library/tests/test__pydantic_settings.py#L109) and when we create an instance in [comp_task.py](https://github.com/ITISFoundation/osparc-simcore/blob/dab37fd0c94e56679f2ce1d9d97f61cd538df09f/services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py#L87) it was highlighted as error

**Review by sanderegg — APPROVED** [review]
pair reviewed

File: packages/models-library/tests/test__models_examples.py (hunk 5 above)
Comment (pcrespov) [reply packages/models-library/tests/test__models_examples.py:46]: As discussed, it will be incorporated on a last stage

File: services/director-v2/src/simcore_service_director_v2/modules/dask_client.py (hunk 10 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/modules/dask_client.py:314]: in place

**Review by KZzizzle — APPROVED** [review]
codeowner approval

File: packages/models-library/src/models_library/service_settings_labels.py (hunk 4 above)
Comment (pcrespov) [reply packages/models-library/src/models_library/service_settings_labels.py:73]: discussed
